src/core/mac/sub_mac_wed.cpp

```diff
@@ -0,0 +1,89 @@
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file implements the Wake-up End Device of the subset of IEEE 802.15.4 MAC primitives.
+ */
+
+#include "sub_mac.hpp"
+
+#if OPENTHREAD_CONFIG_WAKEUP_END_DEVICE_ENABLE
+
+#include "instance/instance.hpp"
+
+namespace ot {
+namespace Mac {
+
+RegisterLogModule("SubMac");
+
+void SubMac::WedInit(void)
+{
+    mWedListenInterval = 0;
+    mWedTimer.Stop();
+}
+
+void SubMac::UpdateWakeupListening(uint32_t aInterval, uint16_t aDuration, uint8_t aChannel)
+{
+    OT_ASSERT(RadioSupportsReceiveTiming());
+
+    VerifyOrExit(aInterval != mWedListenInterval || aDuration != mWedListenDuration || aChannel != mWakeupChannel);
+
+    mWedListenInterval = aInterval;
+    mWedListenDuration = aDuration;
+    mWakeupChannel     = aChannel;
+    mWedTimer.Stop();
+
+    if (mWedListenInterval > 0)
+    {
+        mWedSampleTime      = TimerMicro::GetNow() + kCslReceiveTimeAhead;
+        mWedSampleTimeRadio = TimeMicro((uint32_t)otPlatRadioGetNow(&GetInstance()) + kCslReceiveTimeAhead);
+        HandleWedTimer();
+    }
+
+exit:
+    return;
+}
+
+void SubMac::HandleWedTimer(Timer &aTimer) { aTimer.Get<SubMac>().HandleWedTimer(); }
+
+void SubMac::HandleWedTimer(void)
+{
+    if (RadioSupportsReceiveTiming() && (mState != kStateDisabled))
+    {
+        IgnoreError(Get<Radio>().ReceiveAt(mWakeupChannel, mWedSampleTimeRadio.GetValue(), mWedListenDuration));
```

Comment:
When the timer fires, we are at the scheduled fire time of `mWedSampleTime - kCslReceiveTimeAhead`. We then instruct the radio to `ReceiveAt()` the `mWedSampleTimeRadio`, which should be `kCslReceiveTimeAhead` away from the current time.

Do we need to wait until `kCslReceiveTimeAhead` away from the intended sample time? Can we instruct the radio earlier? `ReceiveAt()` should allow us to schedule receive far in advance.

How about the following?

- We call `ReceiveAt()` setting up the next sample window
- Then schedule the timer for end of the window (start + duration + maybe plus some margin)
- When timer expires, we update the `mWedSampleTime` and do the same

Reply:
What would be the benefit of scheduling so far in advance? Current approach has been working for CSL as well.

Reply:
The main benefit is avoiding a missed receive schedule, especially since we know the intended time and can schedule it earlier.

`kCslReceiveTimeAhead` is not very long (the default value is `#define OPENTHREAD_CONFIG_CSL_RECEIVE_TIME_AHEAD 320`).

I see no advantage in waiting until 320 µs before the scheduled time to call `ReceiveAt()`. Doing so introduces the risk of the radio missing the schedule due to command communication delays.

Also I see no risk with scheduling `ReceiveAt()` early. The beauty of `ReceiveAt()` is that it allows us to schedule the next receive time and be done with it, without requiring the OT code to track and manage exact timings, leaving that responsibility to the radio.

Regarding CSL, we can consider whether it makes sense to change the CSL implementation to schedule `ReceiveAt()` earlier, rather than waiting until 320 µs before the scheduled time. 

Thoughts?

Reply:
I'm a bit afraid of switching to the new approach without thorough testing, but lets follow it for now.

Reply:
I understand the hesitation (to be safe), but frankly, I'm not sure the earlier model of waiting close to the sample time to call `ReceiveAt()` is a good approach, even for CSL. It seems to defeat the purpose of using the `ReceiveAt()` API.

There's an aspect of how `otPlatRadioReceiveAt()` should behave that may be good to clarify. I'll file an issue to discuss this further.

Reply:
I guess platforms that support `ReceiveAt` usually set `kCslReceiveTimeAhead` to a value bigger than 320us to eliminate the risk of missing the slot and hence we assumed this approach to be correct.

For the wake-up sniffing in the current shape and CSL, scheduling RX slots as early as possible sounds like a good idea, but in the case of the enhanced CSL the phase is adjusted according to the CST IE sent by the Wake-up Coordinator, so if a slot is scheduled too early, it may need to be cancelled and scheduled again when a new frame with CST IE has been delivered in the meantime. Thus, it may generate extra radio operations.

But yeah, we can weigh pros and cons in the issue.